File: Java/Visitor/Gemini 1.5 Flash/round1/code.java
import java.util.HashMap;
import java.util.Map;

// Abstract ATM operation interface
interface ATMOperation {
    void execute(Account account);
}

// Concrete ATM operations
class BalanceDisplay implements ATMOperation {
    @Override
    public void execute(Account account) {
        System.out.println("Your current balance is: " + account.getBalance());
    }
}

class CashWithdrawal implements ATMOperation {
    private double amount;

    public CashWithdrawal(double amount) {
        this.amount = amount;
    }

    @Override
    public void execute(Account account) {
        if (account.getBalance() >= amount) {
            account.withdraw(amount);
            System.out.println("Cash withdrawn successfully.");
        } else {
            System.out.println("Insufficient funds.");
        }
    }
}

class Deposit implements ATMOperation {
    private double amount;

    public Deposit(double amount) {
        this.amount = amount;
    }

    @Override
    public void execute(Account account) {
        account.deposit(amount);
        System.out.println("Deposit successful.");
    }
}

class PINChange implements ATMOperation {
    private String newPIN;

    public PINChange(String newPIN) {
        this.newPIN = newPIN;
    }

    @Override
    public void execute(Account account) {
        // Simulate PIN verification (would involve database interaction in real scenario)
        if (newPIN.equals("correct_new_pin")) {
            account.setPIN(newPIN);
            System.out.println("PIN changed successfully.");
        } else {
            System.out.println("PIN change failed. Please try again.");
        }
    }
}

// Account class
class Account {
    private String cardNumber;
    private String PIN;
    private double balance;

    public Account(String cardNumber, String PIN, double balance) {
        this.cardNumber = cardNumber;
        this.PIN = PIN;
        this.balance = balance;
    }

    public String getCardNumber() {
        return cardNumber;
    }

    public String getPIN() {
        return PIN;
    }

    public double getBalance() {
        return balance;
    }

    public void setPIN(String PIN) {
        this.PIN = PIN;
    }

    public void deposit(double amount) {
        balance += amount;
    }

    public void withdraw(double amount) {
        balance -= amount;
    }
}

// ATM visitor interface
interface ATMVisitor {
    void visit(BalanceDisplay operation);
    void visit(CashWithdrawal operation);
    void visit(Deposit operation);
    void visit(PINChange operation);
}

// ATM class
class ATM {
    private Map<String, Account> accounts;

    public ATM() {
        accounts = new HashMap<>();
        // Initialize accounts for testing
        accounts.put("1234567890", new Account("1234567890", "1111", 1000));
    }

    // Simulate card and PIN verification
    public Account authenticate(String cardNumber, String PIN) {
        if (accounts.containsKey(cardNumber) && accounts.get(cardNumber).getPIN().equals(PIN)) {
            return accounts.get(cardNumber);
        } else {
            return null;
        }
    }

    // Accept ATM operation and execute it on the account
    public void processOperation(ATMOperation operation, Account account) {
        operation.execute(account);
    }

    // Accept ATM visitor and let it visit the operations
    public void accept(ATMVisitor visitor) {
        // Example operations
        visitor.visit(new BalanceDisplay());
        visitor.visit(new CashWithdrawal(500));
        visitor.visit(new Deposit(200));
        visitor.visit(new PINChange("new_pin"));
    }
}

// Example ATM visitor implementation
class PrintReceiptVisitor implements ATMVisitor {
    @Override
    public void visit(BalanceDisplay operation) {
        System.out.println("Balance Display operation");
    }

    @Override
    public void visit(CashWithdrawal operation) {
        System.out.println("Cash Withdrawal operation: " + operation.amount);
    }

    @Override
    public void visit(Deposit operation) {
        System.out.println("Deposit operation: " + operation.amount);
    }

    @Override
    public void visit(PINChange operation) {
        System.out.println("PIN Change operation");
    }
}

public class ATMSystem {
    public static void main(String[] args) {
        ATM atm = new ATM();

        // Simulate user authentication
        Account account = atm.authenticate("1234567890", "1111");
        if (account != null) {
            System.out.println("Authentication successful.");

            // Process operations
            atm.processOperation(new BalanceDisplay(), account);
            atm.processOperation(new CashWithdrawal(200), account);
            atm.processOperation(new Deposit(100), account);
            atm.processOperation(new PINChange("new_pin"), account);

            // Print receipt using Visitor pattern
            atm.accept(new PrintReceiptVisitor());
        } else {
            System.out.println("Authentication failed.");
        }
    }
}

File: Python/Visitor/Gemini 1.5 Flash/round1/code.py
from abc import ABC, abstractmethod
from typing import List, Tuple, Dict

class ATMCard:
    def __init__(self, card_number: str, pin: str, balance: float):
        self.card_number = card_number
        self.pin = pin
        self.balance = balance

    def get_balance(self) -> float:
        return self.balance

    def update_balance(self, amount: float):
        self.balance += amount

    def change_pin(self, new_pin: str):
        self.pin = new_pin

class ATMVisitor(ABC):
    @abstractmethod
    def visit_balance_display(self, atm_card: ATMCard):
        pass

    @abstractmethod
    def visit_cash_withdrawal(self, atm_card: ATMCard, amount: float):
        pass

    @abstractmethod
    def visit_deposit(self, atm_card: ATMCard, amount: float):
        pass

    @abstractmethod
    def visit_pin_change(self, atm_card: ATMCard, new_pin: str):
        pass

class BalanceDisplayVisitor(ATMVisitor):
    def visit_balance_display(self, atm_card: ATMCard):
        print(f"Your current balance is: {atm_card.get_balance()}")

    def visit_cash_withdrawal(self, atm_card: ATMCard, amount: float):
        pass

    def visit_deposit(self, atm_card: ATMCard, amount: float):
        pass

    def visit_pin_change(self, atm_card: ATMCard, new_pin: str):
        pass

class CashWithdrawalVisitor(ATMVisitor):
    def visit_balance_display(self, atm_card: ATMCard):
        pass

    def visit_cash_withdrawal(self, atm_card: ATMCard, amount: float):
        if atm_card.get_balance() >= amount:
            atm_card.update_balance(-amount)
            print(f"Cash withdrawn: {amount}. Your new balance is: {atm_card.get_balance()}")
        else:
            print("Insufficient funds.")

    def visit_deposit(self, atm_card: ATMCard, amount: float):
        pass

    def visit_pin_change(self, atm_card: ATMCard, new_pin: str):
        pass

class DepositVisitor(ATMVisitor):
    def visit_balance_display(self, atm_card: ATMCard):
        pass

    def visit_cash_withdrawal(self, atm_card: ATMCard, amount: float):
        pass

    def visit_deposit(self, atm_card: ATMCard, amount: float):
        atm_card.update_balance(amount)
        print(f"Deposited amount: {amount}. Your new balance is: {atm_card.get_balance()}")

    def visit_pin_change(self, atm_card: ATMCard, new_pin: str):
        pass

class PINChangeVisitor(ATMVisitor):
    def visit_balance_display(self, atm_card: ATMCard):
        pass

    def visit_cash_withdrawal(self, atm_card: ATMCard, amount: float):
        pass

    def visit_deposit(self, atm_card: ATMCard, amount: float):
        pass

    def visit_pin_change(self, atm_card: ATMCard, new_pin: str):
        # Assuming a basic pin confirmation system for demonstration
        confirm_pin = input("Confirm new PIN: ")
        if new_pin == confirm_pin:
            atm_card.change_pin(new_pin)
            print("PIN changed successfully.")
        else:
            print("PIN confirmation failed.")

class ATM:
    def __init__(self, card_data: Dict[str, Tuple[str, float]]):
        self.card_data = card_data
        self.max_pin_attempts = 3

    def authenticate(self) -> ATMCard:
        card_number = input("Enter your card number: ")
        if card_number not in self.card_data:
            print("Invalid card number.")
            return None

        pin = input("Enter your PIN: ")
        correct_pin = self.card_data[card_number][0]
        attempts = 0
        while attempts < self.max_pin_attempts:
            if pin == correct_pin:
                return ATMCard(card_number, correct_pin, self.card_data[card_number][1])
            else:
                attempts += 1
                pin = input("Incorrect PIN. Please try again: ")
        print("Card retained due to too many incorrect PIN attempts.")
        return None

    def process_transaction(self, atm_card: ATMCard, visitor: ATMVisitor):
        visitor.visit_balance_display(atm_card)
        visitor.visit_cash_withdrawal(atm_card, 0) # Placeholder for potential withdrawal
        visitor.visit_deposit(atm_card, 0) # Placeholder for potential deposit
        visitor.visit_pin_change(atm_card, "") # Placeholder for potential PIN change

    def run(self):
        atm_card = self.authenticate()
        if atm_card:
            while True:
                print("\nChoose an option:")
                print("1. Balance Display")
                print("2. Cash Withdrawal")
                print("3. Deposit")
                print("4. Change PIN")
                print("5. Exit")

                choice = input("Enter your choice: ")

                if choice == '1':
                    self.process_transaction(atm_card, BalanceDisplayVisitor())
                elif choice == '2':
                    amount = float(input("Enter withdrawal amount: "))
                    self.process_transaction(atm_card, CashWithdrawalVisitor())
                elif choice == '3':
                    amount = float(input("Enter deposit amount: "))
                    self.process_transaction(atm_card, DepositVisitor())
                elif choice == '4':
                    new_pin = input("Enter new PIN: ")
                    self.process_transaction(atm_card, PINChangeVisitor())
                elif choice == '5':
                    print("Thank you for using the ATM.")
                    break
                else:
                    print("Invalid choice.")

if __name__ == "__main__":
    # Sample card data: card number, pin, balance
    card_data = {
        "1234567890123456": ("1111", 1000.00),
        "9876543210123456": ("2222", 500.00)
    }
    atm = ATM(card_data)
    atm.run()